File: examples/runtimes/ruby/lib/node_handler.rb
# frozen_string_literal: true

module Blok
  # NodeHandler is the base class that all Blok nodes must inherit from.
  #
  # Subclasses must implement the +execute+ method to define the node's behavior.
  #
  # @example
  #   class MyNode < Blok::NodeHandler
  #     def execute(ctx, config)
  #       name = ctx.request.body&.dig("name") || "World"
  #       { "message" => "Hello, #{name}!" }
  #     end
  #   end
  #
  class NodeHandler
    # Execute the node logic.
    #
    # @param ctx [Blok::Context] the workflow execution context
    # @param config [Hash] the node-specific configuration
    # @return [Object] the result data to be returned in the execution result
    # @raise [NotImplementedError] if the subclass does not implement this method
    def execute(ctx, config)
      raise NotImplementedError, "#{self.class}#execute must be implemented by the subclass"
    end
  end
end

File: examples/runtimes/php/src/NodeHandler.php
<?php

declare(strict_types=1);

namespace Blok;

/**
 * NodeHandler is the interface that all Blok nodes must implement.
 *
 * Each node receives the workflow execution context and a configuration map,
 * and returns the result data or throws an exception on failure.
 */
interface NodeHandler
{
    /**
     * Execute the node logic.
     *
     * @param Context $ctx    The workflow execution context.
     * @param array   $config Node-specific configuration from the workflow definition.
     * @return mixed The result data to be returned to the Blok runner.
     * @throws \Exception If the node execution fails.
     */
    public function execute(Context $ctx, array $config): mixed;
}
